### ISSUE TITLE
Configuring provider using another resource's outputs does not work

### ISSUE THREAD

**Issue by amitkatyal** [hashicorp/terraform-provider-helm#809]
### Terraform, Provider, Kubernetes, and Helm Versions

```
Terraform version: Terraform v1.0.11
Provider version: (hashicorp/helm v2.4.1)
```

### Affected Resource(s)
- helm_release

### Terraform Configuration Files
```
Root Terraform module
 
provider "helm" {
  alias = "helm_release"
  kubernetes {
    host                   = data.aws_eks_cluster.default.endpoint
    cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority.0.data)
    token                  = data.aws_eks_cluster_auth.default.token
  }
}

data "aws_eks_cluster_auth" "default" {
  name = "${var.cluster_name}-cluster"
  depends_on = [module.cluster]
}

data "aws_eks_cluster" "default" {
  name = "${var.cluster_name}-cluster"
  depends_on = [module.cluster]
}

module "cluster" {
  source = "./modules/cluster"

  name = "${var.cluster_name}-cluster"

  vpc_config = module.vpc.config
  iam_config = module.iam.config

  depends_on = [
    module.iam,
    module.vpc,
  ]

  aws_auth_role_map            = var.aws_auth_role_map
  aws_auth_user_map            = var.aws_auth_user_map
  endpoint_public_access       = var.endpoint_public_access
  endpoint_public_access_cidrs = var.endpoint_public_access_cidrs
}

module "helm" {
  source = "./modules/helm"
  cluster_config = module.cluster.config
  cluster_autoscaler_iam_role_arn = module.cluster.cluster_autoscaler_iam_role_arn
  depends_on = [
    data.aws_eks_cluster.default, data.aws_eks_cluster_auth.default
  ]
  providers = {
    helm = helm.helm_release
  }
}

Helm Module

Provider.tf
terraform {
  required_version = ">= 0.13"

  required_providers {
    helm = {
      source  = "hashicorp/helm"
      version = ">=2.4.1"
    }
  }
}
```

### Steps to Reproduce
1. `terraform apply`

### Expected Behavior
helm_release should get deployed to the EKS cluster and on terraform destroy the resources should get destroyed 

### Actual Behavior
The behavior is not consistent. The resources are getting deployed but "terraform destroy" is failing with the below error.
Sometimes it starts failing in the subsequent "terraform apply".

**Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable**

Am I doing something wrong in passing the provider credentials to the helm module? 

It seems there is an issue if the provider configuration is depending on the output of another resource. As per the documentation, the "issue lies with the order in which Terraform itself evaluates the provider blocks vs. actual resources".

https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs

It is recommended that the most reliable way to configure the Kubernetes provider is to ensure that the cluster itself and the Kubernetes provider resources can be managed with separate apply operations. Data sources can be used to convey values between the two stages as needed

Are we supposed to have two separate sets of scripts to deploy the cluster and the Kubernetes resources to the cluster?. Any workaround?

### Community Note
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jbench** [comment]
I'm also experiencing this issue, and _I think_ (not 100% sure) what resolved it for me was to add the `config_context` argument, and set it to default. I also had to destroy then re-create the cluster to get this working.

```
provider "helm" {
  kubernetes {
    config_context         = "default"
    host                   = local.kube_config.host
    cluster_ca_certificate = local.kube_configca_data
    token                  = local.kube_config.token
  }
}
```

Also noting that it worked fine on my local without specifying `config_context`, it was only broken in Jenkins.

**Comment by mconigliaro** [comment]
`config_context = "default"` didn't work for me.

**Comment by AndreasBergmeier6176** [comment]
It works for me as long as I add `config_path = "~/.kube/config"`. Am trying to find what this provider does differently than the Kubernetes provider (where this works using the same arguments).

**Comment by AndreasBergmeier6176** [comment]
So I took the implementation from Kubernetes provider and added it to `newKubeConfig`. Still the same behavior.
I then added a `.ClientConfig()` call to rapidly note a problem in the config and lo and behold - a error for the config was reported.
I then dumped both `ClientConfig` of both Kubernetes and Helm provider and there again I saw that the former **does not** load the config while the latter **does**. While also ensuring that `configPaths` has `len` `0` in both providers.

The only thing that is suspicious now is that `client-go` version of Kubernetes is `0.23` while the one for Helm is `0.24`.
